packages/graphql-mini-transforms/src/webpack.ts

```diff
@@ -3,33 +3,59 @@ import {dirname} from 'path';
 import {loader} from 'webpack';
 import {parse, DocumentNode} from 'graphql';
 import {getOptions} from 'loader-utils';
+import validateOptions from 'schema-utils';
 
-import {cleanDocument, extractImports, toSimpleDocument} from './document';
+import {
+  cleanDocument,
+  extractImports,
+  toSimpleDocument,
+  CleanDocumentOptions,
+} from './document';
 
 interface Options {
+  generateId?: (normalizedSource: string) => string;
   simple?: boolean;
 }
 
+const schema = {
+  type: 'object' as const,
+  properties: {
+    simple: {
+      type: 'boolean' as const,
+    },
+    generateId: {
+      instanceof: 'Function' as const,
+    },
+  },
+};
+
 export default async function graphQLLoader(
   this: loader.LoaderContext,
   source: string | Buffer,
 ) {
   this.cacheable();
 
   const done = this.async();
-  const {simple = false} = getOptions(this) as Options;
+  const options = {simple: false, ...getOptions(this)} as Options;
+
+  validateOptions(schema, options, {name: '@shopify/graphql-mini-transforms'});
 
   if (done == null) {
     throw new Error(
       '@shopify/graphql-loader does not support synchronous processing',
     );
   }
 
+  const cleanDocumentOptions = {
+    generateId: options.generateId,
+  } as CleanDocumentOptions;
```

Comment:
I think you're right. I wasn't very familiar with TypeScript at the time of this PR, so there's bound to be stuff that's not as clean as it could be. 🙂 Using the suggested change.